### PR TITLE
8296440: Remove Method* handling from cleanup_inline_caches_impl

### PR DESCRIPTION
**Pull request by reinrich**
This PR replaces cleaning of static stubs in CompiledMethod::cleanup_inline_caches_impl() with a guarantee that it is actually not needed because the holder of the embedded target Method* is alive if the caller nmethod is not unloading.

The holder of the target Method* has to be alive because it is reachable from the caller nmethod's oop pool. This is checked by `check_path_to_callee()` when a statically bound call gets resolved.

C2i entry barriers can be removed for the same reason.

Testing:

Many rounds in our CI testing which includes most JCK and JTREG tests, Renaissance benchmark and SAP specific tests with fastdebug and release builds on the standard platforms plus PPC64.

I've also done tier1 and tier2 tests with -XX:-Inline and tier1 tests with ZGC.

I've started hotspot and jdk tier1 tests with -Xcomp. They were not finished when I stopped them after 24h.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296440](https://bugs.openjdk.org/browse/JDK-8296440): Remove Method* handling from cleanup_inline_caches_impl

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [31bcdb07](https://git.openjdk.org/jdk/pull/12802/files/31bcdb07514fefe7bf8d31fb253fed7ed2639401)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [31bcdb07](https://git.openjdk.org/jdk/pull/12802/files/31bcdb07514fefe7bf8d31fb253fed7ed2639401)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [31bcdb07](https://git.openjdk.org/jdk/pull/12802/files/31bcdb07514fefe7bf8d31fb253fed7ed2639401)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12802/head:pull/12802` \
`$ git checkout pull/12802`

Update a local copy of the PR: \
`$ git checkout pull/12802` \
`$ git pull https://git.openjdk.org/jdk.git pull/12802/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12802`

View PR using the GUI difftool: \
`$ git pr show -t 12802`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12802.diff">https://git.openjdk.org/jdk/pull/12802.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12802#issuecomment-1449925101)

### CODE REVIEW

**Review by TheRealMDoerr — APPROVED** [review]
Impressive that we can get rid of so much complicated code! Looks correct to me. Is the check fast enough to do it on every resolution?

**Comment by reinrich** [comment]
> Impressive that we can get rid of so much complicated code! Looks correct to me. 

Thanks for reviewing, Martin.

> Is the check fast enough to do it on every resolution?

I though so. The easy cases where the callee is from a permanent loader or where caller and callee have the same holder oop should be common.
I'm not sure how to quantify the performance effect though.
I wouldn't object to turn it into an assertion if requested.

**Comment by TheRealMDoerr** [comment]
I guess that the benchmarks don't show any impact. If so, I'm ok with your version. If there are concerns, it would be possible to make it switchable by a diagnostic flag. But I'd still keep it enabled by default to get maximum test coverage.

**Review by vnkozlov — COMMENTED** [review]
Let me test it.

I think `check_path_to_callee` should be done only in debug VM. It is verification code.
But for **pre-integration** testing it is fine to have it as guarantee.

**Comment by vnkozlov** [comment]
I hit guarantee in tier7 running our closed tests with JRuby and `-Xcomp -ea -esa -XX:CompileThreshold=100` flags:
```
#  Internal Error (/workspace/open/src/hotspot/share/runtime/sharedRuntime.cpp:1411), pid=1150904, tid=1150907
#  guarantee(false) failed: Missing dependency resolving optimized virtual (invokeinterface) call to  jnr.enxio.channels.Native$LibC$jnr$ffi$2::read
```
Unfortunately I can't share tests.

**Comment by vnkozlov** [comment]
It was linux-x64-debug VM:
```
# JRE version: Java(TM) SE Runtime Environment (21.0) (fastdebug build 21-internal-LTS-2023-03-15-1549029.vladimir.kozlov.jdkgit2)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 21-internal-LTS-2023-03-15-1549029.vladimir.kozlov.jdkgit2, compiled mode, sharing, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x18cca1f]  SharedRuntime::resolve_sub_helper(bool, bool, JavaThread*)+0x96f
```

**Comment by vnkozlov** [comment]
Finally got hs_err file. The top of call stack look like this:
```
V  [libjvm.so+0x18cca1f]  SharedRuntime::resolve_sub_helper(bool, bool, JavaThread*)+0x96f  (sharedRuntime.cpp:1411)
V  [libjvm.so+0x18ccdb9]  SharedRuntime::resolve_helper(bool, bool, JavaThread*)+0x39  (sharedRuntime.cpp:1246)
V  [libjvm.so+0x18cdaee]  SharedRuntime::resolve_opt_virtual_call_C(JavaThread*)+0x15e  (sharedRuntime.cpp:1690)
v  ~RuntimeStub::resolve_opt_virtual_call 0x00007fdab0ddb89b
J 17706 c2 jnr.enxio.channels.Native.read(ILjava/nio/ByteBuffer;)I org.jruby.dist (80 bytes) @ 0x00007fdab1317d9c [0x00007fdab1317d20+0x000000000000007c]
J 17705 c2 jnr.enxio.channels.NativeDeviceChannel.read(Ljava/nio/ByteBuffer;)I org.jruby.dist (91 bytes) @ 0x00007fdab2037704 [0x00007fdab20376c0+0x0000000000000044]
J 17384 c2 org.jruby.util.io.PosixShim.read(Lorg/jruby/util/io/ChannelFD;[BIIZ)I org.jruby.dist (192 bytes) @ 0x00007fdab190984c [0x00007fdab1909720+0x000000000000012c]
J 17379 c2 org.jruby.util.io.OpenFile$2.run(Lorg/jruby/runtime/ThreadContext;Lorg/jruby/util/io/OpenFile;[BII)I org.jruby.dist (82 bytes) @ 0x00007fdab1fff4e8 [0x00007fdab1fff400+0x00000000000000e8]
```

**Comment by vnkozlov** [comment]
> Have you seen this also on aarch64?

I just ran it on linux-aarch64 (Ampere) with the same flags and got the same failure with almost same call stack (without `jnr.enxio`):
```
V  [libjvm.so+0x16c57cc]  SharedRuntime::resolve_sub_helper(bool, bool, JavaThread*)+0x9b8  (sharedRuntime.cpp:1411)
V  [libjvm.so+0x16c595c]  SharedRuntime::resolve_helper(bool, bool, JavaThread*)+0x3c  (sharedRuntime.cpp:1246)
V  [libjvm.so+0x16c6644]  SharedRuntime::resolve_opt_virtual_call_C(JavaThread*)+0x174  (sharedRuntime.cpp:1690)
v  ~RuntimeStub::resolve_opt_virtual_call 0x0000ffff742a37d0
J 32008 c2 org.jruby.util.io.OpenFile$2.run(Lorg/jruby/runtime/ThreadContext;Ljava/lang/Object;[BII)I org.jruby.dist (15 bytes) @ 0x0000ffff769502f8 [0x0000ffff76950280+0x0000000000000078]
```

**Comment by vnkozlov** [comment]
On aarch64 `guarantee` line is the same. I assume `jnr.enxio.channels.Native` methods were inlined that is why they not on call stack:
```
  guarantee(false) failed: Missing dependency resolving optimized virtual (invokeinterface) call to  jnr.enxio.channels.Native$LibC$jnr$ffi$2::read
```

**Comment by reinrich** [comment]
Seems to be independen of `relocInfo::mustIterateImmediateOopsInCode()` then because on aarch64 it is false. Still this is a good point. The verification code should better use `nmethod::oops_do()`.

**Comment by reinrich** [comment]
Reproducer: `./bin/jruby -J-Xcomp -J-ea -J-esa -J-XX:CompileThreshold=100 -S rake spec:ruby:fast
`

**Comment by reinrich** [comment]
> I hit guarantee in tier7 running our closed tests with JRuby and `-Xcomp -ea -esa -XX:CompileThreshold=100` flags:
> 
> ```
> #  Internal Error (/workspace/open/src/hotspot/share/runtime/sharedRuntime.cpp:1411), pid=1150904, tid=1150907
> #  guarantee(false) failed: Missing dependency resolving optimized virtual (invokeinterface) call to  jnr.enxio.channels.Native$LibC$jnr$ffi$2::read
> ```
> 
> Unfortunately I can't share tests.

Thanks again for the testing.

The guarantee fails for the read call at https://github.com/jnr/jnr-enxio/blob/4e21f3a341c2cb7e8007b7bebfb964182d36318d/src/main/java/jnr/enxio/channels/Native.java#L120. It is an invokeinterface with a [constant receiver](https://github.com/jnr/jnr-enxio/blob/4e21f3a341c2cb7e8007b7bebfb964182d36318d/src/main/java/jnr/enxio/channels/Native.java#L67).

I've implemented and added a test. The guarantee fails when CHA based optimization is not possible. Then no dependencies are emitted but the call is still optimized based on the constness of the receiver. Btw wouldn't it be better to skip the CHA attempt in order to avoid emitting dependencies? It might be even possible that the caller gets deoptimized because of them.

I've adapted the verification code. If the receiver is found among the constants then we know a permanent path to the target exists even if the receiver is not constant as the target does not depend on the receiver.

My reproducer `./bin/jruby -J-Xcomp -J-ea -J-esa -J-XX:CompileThreshold=100 -S rake spec:ruby:fast` does not hit the guarantee with the changes anymore.

File: src/hotspot/share/runtime/sharedRuntime.cpp [hunk 1]
```diff
@@ -1351,6 +1351,97 @@ bool SharedRuntime::resolve_sub_helper_internal(methodHandle callee_method, cons
...
+    } else {
+      oop to = to_cld->class_loader();
+      oop from = from_cld->class_loader();
+
+      if (from == to ||
+          (!from_cld->is_boot_class_loader_data() &&
+           java_lang_ClassLoader::isAncestor(from, to))) {
+        return; // `to` is reachable by iterating parents of `from`
+      }
+    }
```
Comment (coleenp) [inline src/hotspot/share/runtime/sharedRuntime.cpp:1418]: I'd be happier if this part was a function in ClassLoaderData or refactored from record_dependency (along with the constains function).  Since it's similar code.

**Review by coleenp — COMMENTED** [review]
I don't really know if this is right but some comments anyway.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 2]
```diff
@@ -4303,11 +4303,6 @@ void MacroAssembler::cmpoop(Register obj1, Register obj2) {
   cmp(obj1, obj2);
 }
 
-void MacroAssembler::load_method_holder_cld(Register rresult, Register rmethod) {
-  load_method_holder(rresult, rmethod);
-  ldr(rresult, Address(rresult, InstanceKlass::class_loader_data_offset()));
```
[the remark this answers is not in the document]
Comment (reinrich) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:4308]: Done

File: src/hotspot/share/runtime/sharedRuntime.cpp [hunk 3]
```diff
@@ -1351,6 +1351,97 @@ bool SharedRuntime::resolve_sub_helper_internal(methodHandle callee_method, cons
   return true;
 }
 
+class Search2OopsClosure : public OopClosure {
```
[the remark this answers is not in the document]
Comment (reinrich) [reply src/hotspot/share/runtime/sharedRuntime.cpp:1354]: Yes. I'll do that before the pr gets integrated.
I want to make sure all dependencies are present also in the release build.
(E.g. before JDK-8299155 was fixed the debug vm produced more dependencies)

File: src/hotspot/share/runtime/sharedRuntime.cpp (hunk 1 above)
Comment (reinrich) [reply src/hotspot/share/runtime/sharedRuntime.cpp:1404]: You are right. I've done the refactoring. Please let me know what you think.

File: src/hotspot/share/classfile/classLoaderData.cpp [hunk 4]
```diff
@@ -787,6 +787,10 @@ ClassLoaderMetaspace* ClassLoaderData::metaspace_non_null() {
   return metaspace;
 }
 
+bool ClassLoaderData::handles_contain(oop obj) {
+  return _handles.contains(obj);
```
[the remark this answers is not in the document]
Comment (reinrich) [reply src/hotspot/share/classfile/classLoaderData.cpp]: I though it was not needed.
Looks like `ChunkedHandleList` supports unsynchronized reading. We might not see all the handles recently added without locking but we are guaranteed to see the handles that were added by resolving the call we are checking.

**Comment by reinrich** [comment]
> I don't really know if this is right but some comments anyway.

Thanks for looking at this and for giving feedback.

**Comment by reinrich** [comment]
Going back to draft while analyzing test errors. Sorry for the noise.

**Comment by reinrich** [comment]
The tests failed because `java_lang_ClassLoader::isAncestor()` must not be called for the boot loader. Protection against this got lost with the refactoring of the last change (https://github.com/openjdk/jdk/pull/12802/commits/2c2e49b5a01cd58d6f922ce105964c083ee18bfb)

So while the boot loader never needs an explicit dependency to the holder of an ordinary class (afaik) because it never delegates loading, it does have explicit dependencies to classes with class mirror holder (e.g. java.lang.invoke.LambdaForm$MH/0x0000000800082800). Therefore checking CLD dependencies cannot be skipped in `check_path_to_callee()` if `from_cld` should be the boot loader cld. Instead the `isAncestor()` call is only done if `from_cld` isn't the boot loader. `isAncestor()` could also be changed to return false if the first arg is the boot loader by simply converting the do-while-loop into a while-loop but then it wouldn't be a verbatim copy from ClassLoader.java anymore.

**Comment by reinrich** [comment]
One test failed in GHA. This is hardly related to these changes:

```
TEST: jdk/internal/misc/ThreadFlock/ThreadFlockTest.java#platform
[...]
org.opentest4j.AssertionFailedError: Duration 5556ms, expected <= 4000ms ==> expected: <true> but was: <false>
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:40)
	at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:210)
	at ThreadFlockTest.checkDuration(ThreadFlockTest.java:1090)
	at ThreadFlockTest.testAwaitAllWithTimeout1(ThreadFlockTest.java:431)
[...]
```

**Comment by reinrich** [comment]
No issues in test.

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the updates! I think it's good for another spin.

**Comment by vnkozlov** [comment]
Started new testing.

**Review by vnkozlov — APPROVED** [review]
No new failures. Tier7 is clean.

**Comment by reinrich** [comment]
> No new failures. Tier7 is clean.

Thanks! Are there tests still pending?

File: src/hotspot/share/runtime/sharedRuntime.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/runtime/sharedRuntime.cpp:1428]: I like the refactoring.  Thank you.

**Review by coleenp — APPROVED** [review]
The ClassLoaderData parts look good to me.

File: src/hotspot/share/classfile/classLoaderData.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/classfile/classLoaderData.cpp]: I think you're right that it isn't needed.

**Comment by reinrich** [comment]
> All tier1-7 finished. Also Xcomp and stress. That is why I approved changes.

Good. Sorry, I've overlooked your approval.
Thanks, Richard.

**Comment by reinrich** [comment]
So from my point of view this is ready for integration. Thanks for the help, feeback, and reviews!

Before I'd like to share some final thoughts. So the search in `check_path_to_callee()` is shallow. It might yield false positives. I hope for support by compiler folks adding dependencies at compile time, even if only needed to avoid false positives, and even more so, if no path exists.

To make this really robust, new dependencies could be added to the caller nmethod if no path to the holder of the call target is found. This is not possible atm but maybe the nmethod's oop pool could be changed to be a growable kind of (weak) OopStorage outside the code cache if needed.

An adhoc fix in the case of missing(?) path could be not to resolve the call if the callee should be interpreted. Instead the callee could be compiled in the background. Execution would continue without resolving the call.

@fisk may I ask if you are also ok with the general approach of this pr?

**Comment by reinrich** [comment]
> I'm okay with the general approach. I'm a bit worried about it though.

I'm also not perfectly comfortable with it. Now even more so.

> With this approach, correct unloading of statically bound callsites relies on the dependency data used for deopt due to class loading, also implicitly embed enough information to deal with the unloading. Is there always a match? In order to answer that question we have to either walking all combinations of paths in which our compilers can generate a monomorphic callsite and see that it's correct, hope it doesn't break over time, and hope that the assert catches when it does, or we will end up crashing very intermittently in very obscure ways.

We might crash intermittently but not in obscure ways. We would hit the guarantee I left in CompiledMethod::cleanup_inline_caches_impl(). It would be clear that the statically bound call lacks the dependency to the target holder. IMHO `check_path_to_callee()` would be cheap enough to have it in release builds at least when resolving an optimized call to an interpreted callee (I'd consider static calls to be always reachable through CLD handles). This would make the crashes less intermittent.
It would be obscure though why the dependency is missing. Probably you meant that.

I agree it's really hard to reason about the paths in the compilers when they emit optimized virtual calls. I don't want to do it. I though I could establish the requirement that the call target dependency must exist and that it is easy to discover. It would be the responsibility of compiler providers to meet the requirement. But I understand now that this is not robust enough.

> I have spent a fair amount of time looking through the different paths in which you get to the decision point of generating a statically bound callsite, and I think that it is currently correct. But there are so many paths, I find it hard to follow them all.
> I would have preferred separating implicit dependency on all the reasons why a callsite became statically bound, with how to deal with correct unloading, so that it's easier to reason about it with certainty. For example, making sure there is an oop slot for each statically bound callsite, and when resolving the callsite, just ensure the holder oop is in there and register_nmethod() to the GC.

Ok, so the relocation for optimized virtual calls should point to an oop slot, right?

> We already have support for calling register_nmethod multiple times due to C1 mirror patching that does a similar trick.

I see. This is needed if, e.g., the patched oop is in the young region.

> Then we could algorithmically ensure that the logic works as expected, regardless of why the callsite became statically bound.
> Having said that, I'm okay with this approach too. We will probably run into the assert if something fishy happens, eventually. So I'll leave it to you to decide, and I'm okay with integrating as is.

So right now I don't feel like integrating the pr as it is.
I thought the approach was ok. In tests the call target holder is always reachable. I assumed we could even add it to CLD::_handles for the sake of robustness if the path is not found. It might look like a hack but given that in reality it is never done it still appears acceptable to me.

Anyways, thanks a lot for bearing that long with me.
Probably you're right. Every optimized call should refer to an oop slot for referencing the target.

**Comment by dean-long** [comment]
BTW, I experimented with a growable nmethod oops table in JDK-8294002, but ran into problems for reasons I can't remember.  I think reserving a slot per call site would be the safest approach.  However I saw a potential problem with platforms like Aarch64 that appear to store the metadata in two places, the move instruction sequence itself and the relocation metadata slot.  If those two places aren't kept consistent then there could be a situation where the call site oop protects one of the values but not the other.

**Comment by reinrich** [comment]
> BTW, I experimented with a growable nmethod oops table in JDK-8294002, but ran into problems for reasons I can't remember. I think reserving a slot per call site would be the safest approach. However I saw a potential problem with platforms like Aarch64 that appear to store the metadata in two places, the move instruction sequence itself and the relocation metadata slot. If those two places aren't kept consistent then there could be a situation where the call site oop protects one of the values but not the other.

The problem is present in HEAD. Modification of Metadata references would require fixup if not stored as an immediate operands in the instruction stream as on x86. But this was never implemented (`metadata_Relocation::pd_fix_value()` is empty) since Metadata objects don't move. So if the [target Method* is cleared from a static stub](https://github.com/openjdk/jdk/blob/ecec611af6c6314d7a834392f38468ad3f390e2d/src/hotspot/share/code/compiledMethod.cpp#L661-L665) readers will see this but executing threads would get the dangling pointer because the fixup wasn't done.

If holder of the target is stored in an oop slot associated with the optimized call then cleaning of static stubs can be removed because the holder will be kept alive if the caller is on stack. If it is not on stack and the holder is unloading then the caller will be unloading too.

**Comment by reinrich** [comment]
Closing this PR for now (with the assumption that comments are still possible).
